### PR TITLE
HDDS-7650. Recon - Data Management Metrics - Understand the Age of Data.

### PR DESCRIPTION
**Pull request by devmadhuu**
Understand the age of data ( age of buckets, age of objects, age of  volumes) 
  1. This requirement will be limited by the number of buckets (say top 10) to show with their age information as a given volume can have a large number of buckets. Same holds true for showing a limited number of volumes (say top 10) with their age information if root path  is given. 
  2. Top 10 criteria can be determined based on following factors:
  3. Age of Top 10 buckets can be listed based on either count of keys or size of buckets, if path is upto a specific volume.
  4. Age of top 10 volumes can be listed based on either count of buckets or size of volumes if path is given as root .
  5. If the path given by the user is upto bucket level,  then top 10 keys based on size will be given with their age details.
  6. If the path given by the user is upto key or leaf level, then only that key age details will be returned along with size.

https://issues.apache.org/jira/browse/HDDS-7650

## How was this patch tested?

This patch is tested using postman by verifying backend responses:
API details:
If Path is Bucket:

Keys
TopN Keys with Age, Size details
By Count - Give user option to select 10, 20, 30 etc
By Size - Give user option to select With or Without Replica
Dirs
TopN Dirs with Age, Size, dirCount, keycount details
By Count - Give user option to select 10, 20, 30 etc
By Size - Give user option to select With or Without Replica

If Path is Volume:

Buckets
TopN Buckets with Age, Size, dirCount, keycount details
By Count - Give user option to select 10, 20, 30 etc
By Size - Give user option to select With or Without Replica

If Path is Directory:

Keys
TopN Keys with Age, Size details
By Count - Give user option to select 10, 20, 30 etc
By Size - Give user option to select With or Without Replica
Dirs
TopN Dirs with Age, Size, dirCount, keycount details
By Count - Give user option to select 10, 20, 30 etc
By Size - Give user option to select With or Without Replica

If Path is Key:

Display: Size, Age details

API End Point :
Request Type: GET
API Path: /api/v1/namespace/entitymetrics?path=<path of volume/bucket/dir/key>
Sample Response:
{
"entityInfo":

{ "type": "DIRECTORY", "numVolume": -1, "numBucket": -1, "numDir": 2, "numKey": 5, "status": "OK" }

,
"childEntityMetrics": [

{ "key": false, "path": "/volume1/bucket1/dir1/dir2", "size": 86060, "sizeWithReplica": 258180, "isKey": false, "keyCount": 4, "bucketCount": -1, "dirCount": 1, "volumeCount": -1, "type": "DIRECTORY", "creationTime": 1671801625290, "modificationTime": 1671801625290 }

,

{ "key": true, "path": "/volume1/bucket1/dir1/LICENSE.txt", "size": 21515, "sizeWithReplica": 64545, "isKey": true, "keyCount": 0, "bucketCount": -1, "dirCount": -1, "volumeCount": -1, "type": "KEY", "creationTime": 1671801625290, "modificationTime": 1671801625290 }

],
"status": "OK"
}

### CODE REVIEW

**Comment by devmadhuu** [comment]
@sumitagrawl @dombizita @smengcl , pls review

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/NamespaceSummaryResponse.java [hunk 1]
```diff
@@ -43,6 +43,16 @@ public class NamespaceSummaryResponse {
   @JsonProperty("numKey")
   private long numTotalKey;
 
+  /** Age of entity in
+   * human-readable format. */
+  @JsonProperty("age")
+  private long age;
```
Comment (sumitagrawl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/NamespaceSummaryResponse.java:49]: can have same name as createTime, UI can calculate age and display name accordingly

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java [hunk 2]
```diff
@@ -62,6 +70,129 @@ public NSSummaryEndpoint(ReconNamespaceSummaryManager namespaceSummaryManager,
     this.reconSCM = reconSCM;
   }
 
+  @GET
+  @Path("/entitymetrics")
+  public Response getEntityMetrics(@QueryParam("path") String path,
+                                   @DefaultValue("size")
```
Comment (sumitagrawl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:76]: orderby should include "age" also a filter and as default

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java [hunk 3]
```diff
@@ -62,6 +70,129 @@ public NSSummaryEndpoint(ReconNamespaceSummaryManager namespaceSummaryManager,
...
+          childMetricsListMap.put(EntityType.UNKNOWN.name(),
+              duResponse.getDuData()
+              .stream().sorted(Comparator.comparingLong(
+                  DUResponse.DiskUsage::getSizeWithReplica).reversed())
+              .limit(count).collect(Collectors.toList()));
+          break;
+        }
+      }
+      entityMetricsResponse.setNsSummaryResponse(nsSummaryData);
+      entityMetricsResponse.setChildMetricsListMap(childMetricsListMap);
```
Comment (sumitagrawl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:175]: childMetricsList should be part of NsSummary as it belongs to it.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java [hunk 4]
```diff
@@ -62,6 +70,129 @@ public NSSummaryEndpoint(ReconNamespaceSummaryManager namespaceSummaryManager,
     this.reconSCM = reconSCM;
   }
 
+  @GET
+  @Path("/entitymetrics")
```
Comment (sumitagrawl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:74]: uri name can be entityInfo or other better name, not a metrics as its not a metric data.

**Review by sumitagrawl — CHANGES_REQUESTED** [review]

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/NamespaceSummaryResponse.java (hunk 1 above)
Comment (devmadhuu) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/NamespaceSummaryResponse.java]: @sumitagrawl - Thanks for reviewing. It has been fixed. Pls re-review.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java (hunk 2 above)
Comment (devmadhuu) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:76]: @sumitagrawl - Thanks for reviewing. It has been fixed. Pls re-review.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java (hunk 4 above)
Comment (devmadhuu) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java]: @sumitagrawl - Thanks for reviewing. It has been fixed. Pls re-review.

**Comment by kerneltime** [comment]
@duongkame can you please take a look?

**Comment by devmadhuu** [comment]
@ChenSammi , @duongkame  pls review

**Review by sumitagrawl — COMMENTED** [review]
LGTM, please add Unit cases

**Comment by devmadhuu** [comment]
> LGTM, please add Unit cases

@sumitagrawl  - Added test cases. Pls review.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java [hunk 5]
```diff
@@ -48,6 +56,8 @@
 @AdminOnly
 public class NSSummaryEndpoint {
 
+  public static final String COUNT = "count";
```
Comment (fapifta) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:59]: Shouldn't we have createTime here? Also shouldn't that be cTime or creationTime?
As these three belongs together, wouldn't it worth it to create an enum for the values we accept, and validate and parse the parameters based on the enum?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java [hunk 6]
```diff
@@ -62,6 +72,172 @@ public NSSummaryEndpoint(ReconNamespaceSummaryManager namespaceSummaryManager,
     this.reconSCM = reconSCM;
   }
 
+  @GET
+  @Path("/entityinfo")
+  public Response getEntityMetrics(@QueryParam("path") String path,
```
Comment (fapifta) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:77]: Can we add API documentation for this method if we keep it?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java [hunk 7]
```diff
@@ -116,21 +294,49 @@ public Response getDiskUsage(@QueryParam("path") String path,
     if (path == null || path.length() == 0) {
       return Response.status(Response.Status.BAD_REQUEST).build();
     }
+    return Response.ok(getDuResponse(path, listFile, withReplica)).build();
+  }
 
+  private DUResponse getDuResponse(String path, boolean listFile,
+                                   boolean withReplica) throws IOException {
+    AtomicBoolean isError = new AtomicBoolean(false);
```
Comment (fapifta) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:302]: I believe we can use a simple boolean here, even if the forEach runs in parallel mode, the only change to the value is to set it to true, and we just check after execution is done, so we don't need atomicity on it as it does not matter i one or 3k threads sets it to true at once, but correct me if I am wrong.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java [hunk 8]
```diff
@@ -116,21 +294,49 @@ public Response getDiskUsage(@QueryParam("path") String path,
     if (path == null || path.length() == 0) {
       return Response.status(Response.Status.BAD_REQUEST).build();
     }
+    return Response.ok(getDuResponse(path, listFile, withReplica)).build();
```
Comment (fapifta) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:297]: After extracting the method here, but even before that, we can remove the @SuppressWarnings for the /du endpoint method in line 287

**Review by fapifta — COMMENTED** [review]
I have some doubts with the getEntityMetrics endpoint that we are defining in this patch, and also I am not sure why we need the EntityInfoResponse... So let's discuss this first, depending on the answers I might have some suggestions for the patch itself besides what I noted inline.

Probably this is just my ignorance, but it is unclear where we would like to show this data...

The full dataset that the EntityInfoResponse seem to provide can be calculated from the results of /summary and /du if I'm correct.
If during rendering the UI we call both endpoints that we call from getEntityMetrics as well, then it is pointless to have a separate endboint, as that just pushes calculation burden on Recon, while client side should be able to calculate the data in theory.

If the calculation can not be done on the client side for any reason...
The thing is that EntityInfoRespone, DUResponse and CountStats seems to be redundant, and it feels like they are storing the same information, what is the reason behind these three?
Also the getEntityMetrics method does not use the EntityHandlers, but it has a big if - elseif - else statement within each a fairly similar switch - case statement, should that be done in the EntityHandler implementations which would help to simplify this newly added method as I feel when I look at it. Is there a reason to keep these if-switch blocks?

**Comment by smitajoshi12** [comment]
Can we get mutiple Parent Entity Records?

**Comment by DaveTeng0** [comment]
Hey @devmadhuu ~ there are some new comments. Please help address, thanks!

**Pull request closed without merging**